File: crop_markers.py
import os
from PIL import Image, ImageOps

image_path = "ARmarkers.png"
img = Image.open(image_path).convert("RGB")
img_w, img_h = img.size

# Scan from the bottom to find where content ends (last row with dark pixels)
import numpy as np
pixels = np.array(ImageOps.grayscale(img))
# Find the last row that has a pixel darker than 200
dark_rows = np.where(pixels.min(axis=1) < 200)[0]
content_bottom = int(dark_rows[-1]) + 1 if len(dark_rows) else img_h

print(f"Image size: {img_w}x{img_h}")
print(f"Content ends at row: {content_bottom}")

# Crop to just the grid area
grid = img.crop((0, 0, img_w, content_bottom))
grid_w, grid_h = grid.size

rows, cols = 4, 5
cell_w = grid_w / cols
cell_h = grid_h / rows

labels = [
    ["CELC", "E1", "E1A", "E2", "E2A"],
    ["E3", "E3A", "E4", "E4A", "E5"],
    ["E6", "E8", "EA", "EW1", "EW1A"],
    ["SDE4", "SDE2", "SDE1", "SDE3", "T-Lab"]
]

output_dir = "cropped_qr_codes"
os.makedirs(output_dir, exist_ok=True)

for r in range(rows):
    for c in range(cols):
        left   = c * cell_w
        top    = r * cell_h
        right  = (c + 1) * cell_w
        bottom = (r + 1) * cell_h
        cropped = grid.crop((left, top, right, bottom))
        # Trim white padding — keep from the outermost dark pixel inward
        gray_cell = ImageOps.grayscale(cropped)
        tight_bbox = gray_cell.point(lambda p: 255 if p < 200 else 0).getbbox()
        if tight_bbox:
            cropped = cropped.crop(tight_bbox)
        file_name = f"{labels[r][c]}.png"
        cropped.save(os.path.join(output_dir, file_name))
        print(f"Saved: {file_name}")

print(f"\nDone! {rows * cols} markers saved to '{output_dir}'")

File: crop-markers.js
const puppeteer = require('puppeteer');
const { Jimp } = require('jimp');
const path = require('path');
const fs = require('fs');
const http = require('http');

const PDF_PATH = path.resolve(__dirname, 'ARmarkers.pdf');
const OUT = path.resolve(__dirname, 'cropped_qr_codes');
fs.mkdirSync(OUT, { recursive: true });

const labels = [
  ['CELC', 'E1', 'E1A', 'E2', 'E2A'],
  ['E3', 'E3A', 'E4', 'E4A', 'E5'],
  ['E6', 'E8', 'EA', 'EW1', 'EW1A'],
  ['SDE4', 'SDE2', 'SDE1', 'SDE3', 'T-Lab'],
];
const ROWS = 4, COLS = 5;

const pdfBase64 = fs.readFileSync(PDF_PATH).toString('base64');

// Serve a page that renders the PDF via PDF.js to a canvas
const html = `<!DOCTYPE html><html><head>
<script src="https://cdnjs.cloudflare.com/ajax/libs/pdf.js/3.11.174/pdf.min.js"></script>
</head><body style="margin:0;background:#fff">
<canvas id="c"></canvas>
<script>
pdfjsLib.GlobalWorkerOptions.workerSrc = 'https://cdnjs.cloudflare.com/ajax/libs/pdf.js/3.11.174/pdf.worker.min.js';
const data = atob('${pdfBase64}');
const bytes = new Uint8Array(data.length);
for (let i = 0; i < data.length; i++) bytes[i] = data.charCodeAt(i);
pdfjsLib.getDocument({ data: bytes }).promise.then(pdf => pdf.getPage(1)).then(page => {
  const vp = page.getViewport({ scale: 3 });
  const canvas = document.getElementById('c');
  canvas.width = vp.width; canvas.height = vp.height;
  return page.render({ canvasContext: canvas.getContext('2d'), viewport: vp }).promise;
}).then(() => { document.title = 'done'; }).catch(e => { document.title = 'error:' + e.message; });
</script></body></html>`;

const server = http.createServer((req, res) => {
  res.writeHead(200, { 'Content-Type': 'text/html' });
  res.end(html);
});

server.listen(7790, async () => {
  console.log('Rendering PDF via PDF.js...');
  const browser = await puppeteer.launch({ headless: true, args: ['--no-sandbox'] });
  const page = await browser.newPage();
  await page.goto('http://localhost:7790/', { waitUntil: 'networkidle2', timeout: 30000 });

  await page.waitForFunction(() => document.title === 'done' || document.title.startsWith('error'), { timeout: 30000 });
  const title = await page.title();
  if (title.startsWith('error')) { console.error(title); await browser.close(); server.close(); return; }

  // Get canvas dimensions and screenshot just the canvas
  const dims = await page.evaluate(() => {
    const c = document.getElementById('c');
    return { w: c.width, h: c.height };
  });
  console.log('Rendered PDF canvas:', dims.w, 'x', dims.h);

  await page.setViewport({ width: dims.w, height: dims.h });
  const raw = path.join(__dirname, '_markers_raw.png');
  await page.screenshot({ path: raw });
  await browser.close();
  server.close();

  console.log('Cropping', ROWS * COLS, 'markers...');
  const img = await Jimp.read(raw);
  const W = img.bitmap.width, H = img.bitmap.height;
  const cellW = Math.floor(W / COLS), cellH = Math.floor(H / ROWS);

  for (let r = 0; r < ROWS; r++) {
    for (let c = 0; c < COLS; c++) {
      const label = labels[r][c];
      const cropped = img.clone().crop({ x: c * cellW, y: r * cellH, w: cellW, h: cellH });
      await cropped.write(path.join(OUT, label + '.png'));
      console.log('Saved:', label + '.png');
    }
  }
  console.log('\nDone! Markers saved to', OUT);
});
